### PR TITLE
Adding support for custom models

### PR DESCRIPTION
**Pull request by eric-czech**
Hi @zachmayer ,  here is a PR for custom model support with some more docs and at least one decent test.  Let me know if you have any suggestions.

Also for the sake of reference, here is some example usage:

```
library(plyr); library(dplyr)
library(caret); library(caretEnsemble)

# Generate some data to test with
d <- twoClassSim(n=100)
X <- d %>% select(-Class); y <- d$Class

# Create a couple custom models
customRF <- getModelInfo('rf', regex=F)[[1]]
customRF$method <- 'custom.rf'

customGLM <- getModelInfo('rf', regex=F)[[1]]
customGLM$method <- 'custom.glm'

cl <- caretList(
  X, y,
  tuneList=list(
    # The name used internally for this model will come from the "method" attribute above
    caretModelSpec(method=customRF, tuneLength=3),
    # This model on the other hand will be referred to as "myglm" not "custom.glm"
    myglm=caretModelSpec(method=customGLM, tuneLength=1),
    glmnet=caretModelSpec(method='glmnet', tuneLength=5),
    rpart=caretModelSpec(method='rpart', tuneLength=15)
  ),
  trControl=trainControl(method='cv', number=10, classProbs=T)
)

cs <- caretEnsemble(cl)
print(cs)

# A glm ensemble of 4 base models: custom.rf, myglm, glmnet, rpart
# ...
```

ps I closed the first PR for this after a bad merge on that branch, and had to create this one instead.

### CODE REVIEW

**Comment by zachmayer** [comment]
Looks good to me.  Please fix the lint errors, and I'll do a final review and merge.
Thank you!

**Comment by eric-czech** [comment]
Will do (I'm traveling so sorry for the delay).  I don't know why those lint errors aren't all showing up when running the local tests.  Oh well I'll get them fixed eventually.

**Comment by eric-czech** [comment]
@zachmayer do you have any idea why this build keeps failing?  I can't really tell just based on the TravisCI logs.  I fixed all the lint errors (and I keep squashing those changes down to keep everything in one commit).

That most recent issue with the build mentioned by lintr-bot was definitely fixed so I'm at a loss.

**Comment by zachmayer** [comment]
[Taking a look at the travis logs](https://travis-ci.org/zachmayer/caretEnsemble/jobs/117411068), I see one warning and one note:
```{R}
* checking R code for possible problems ... NOTE
methodCheck: no visible binding for global variable ‘type’
```
...
```{R}
* checking Rd \usage sections ... WARNING
Undocumented arguments in documentation object 'validateCustomModel'
  ‘x’
Documented arguments not in \usage in documentation object 'validateCustomModel':
  ‘a’
```

The warning is causing the test failure— I think you just need to update the `@params` section of the roxygen comment block above your function.

You can run these checks locally with `devtools::test()`, which run's all of CRAN's unit tests, on top of the testthat tests.

File: man/validateCustomModel.Rd [hunk 1]
```diff
@@ -0,0 +1,20 @@
+% Generated by roxygen2: do not edit by hand
+% Please edit documentation in R/helper_functions.R
+\name{validateCustomModel}
+\alias{validateCustomModel}
+\title{Validate a custom caret model info list}
+\usage{
+validateCustomModel(x)
+}
+\arguments{
+\item{a}{model info list (e.g. getModelInfo("rf", regex=F)[[1]])}
```
Comment (zachmayer) [inline man/validateCustomModel.Rd]: This miss-match is what's causing the text failure.  The argument of the function is `x` and the argument you documented is `a`.

HOWEVER, in the roxygen comment block, the documentation looks correct to me.  I think running `devtools::document()` and pushing the result will fix the problem.

File: R/helper_functions.R [hunk 2]
```diff
@@ -195,6 +195,36 @@ check_bestpreds_preds <- function(modelLibrary){
...
+  } else if (x$method == "custom"){
+    validateCustomModel(x$modelInfo)$method
+  } else x$method
+}
+
+#' @title Validate a custom caret model info list
+#' @description Currently, this only ensures that all model info lists
+#' were also assigned a "method" attribute for consistency with usage
+#' of non-custom models
+#' @param x a model info list (e.g. \code{getModelInfo("rf", regex=F)\[[1]]})
```
Comment (zachmayer) [inline R/helper_functions.R]: This looks correct to me, but doesn't match the `.Rd` file (see below).  I think running `devtools::document()` will fix the problem.

You can check your work locally with `devtools::check()`, which conveniently also runs `devtools::document()`.

**Comment by zachmayer** [comment]
I added some comments that I think will help you fix the problem.  Please let me know if you have any other issues, and I'll help you work them out!

**Comment by eric-czech** [comment]
Beautiful, thanks for the assist!

**Comment by zachmayer** [comment]
No problem, it's what I'm here for lol!

**Comment by smead10** [comment]
Hi @eric-czech  -- 

I have come across this error and still having issues. I followed your fix in creating custom models ... 

The error I get is is.character(method) is not true. 
I think I have updated everything including caretEnsamble.  I am on R version 3.3.3. Windows 7 x64 

Any ideas ? 

Much appreciated. 

Cheers,
Sanida

**Pull request merged**